pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+
+        /// Returns a multilocation that can be used in the `deposit_asset` XCM opcode.
+        /// The `interior_multilocation` can be obtained using `Self::interior_multilocation`.
+        pub fn absolute_multilocation(
+            interior_multilocation: InteriorMultiLocation,
+        ) -> MultiLocation {
+            let mut l = interior_multilocation;
+            l.push_front(Junction::Parachain(T::SelfParaId::get().into()))
+                .expect("multilocation too long");
+            MultiLocation::from(l)
+        }
+
+        fn on_collator_instantaneous_core_requested(para_id: ParaId) -> DispatchResult {
+            let mut in_flight_orders = InFlightOrders::<T>::get();
+            if in_flight_orders.contains(&para_id) {
+                return Err(Error::<T>::OrderAlreadyExists.into());
+            }
+            in_flight_orders
+                .try_insert(para_id)
+                .map_err(|_| Error::<T>::InFlightLimitReached)?;
+
+            // Check that the para id is a parathread
+            let parathread_params = T::GetParathreadParams::get_parathread_params(para_id);
+            if parathread_params.is_none() {
+                return Err(Error::<T>::NotAParathread.into());
+            }
+
+            // TODO: also compare the latest slot from pallet_author_noting with parathread_params.slot_frequency
+
+            let xcm_weights_storage =
+                XcmWeights::<T>::get().ok_or(Error::<T>::XcmWeightStorageNotSet)?;
+
+            let withdraw_amount = xcm_weights_storage.buy_execution_cost;
+
+            // Send xcm to the relay
+            // Use a derivative account from the sovereign account based on the paraId
+            // Buy on-demand cores
+            // Any failure should return everything to the derivative account
+
+            // Don't use utility::as_derivative because that will make the tanssi sovereign account
+            // pay for fees, instead use `DescendOrigin` to make the container chain sovereign account
+            // pay for fees. The container chain sovereign account is derived from the tanssi sovereign
+            // account.
+            // TODO: when coretime is implemented, buy core there instead of buying on-demand cores
+            let origin = OriginKind::SovereignAccount;
+            // TODO: max_amount is the max price of a core that this parathread is willing to pay
+            // It should be defined in a storage item somewhere, controllable by the container chain
+            // manager.
+            let max_amount = u128::MAX;
+            let call = T::GetPurchaseCoreCall::get_encoded(max_amount, para_id);
+            let weight_at_most = xcm_weights_storage.weight_at_most;
+
+            // Assumption: derived account already has DOT
+            // The balance should be enough to cover
+            // TODO: we could make this be part of the proof, so collators cannot call this if the
+            // derived account does not have enough balance
```

Comment:
I strongly oppose to this idea, for the reasons that we discussed already. We can check that the balance is correct in the block at which we are builiding, but have no guarantees that when the XCM reaches the relay this is true

Reply:
Ah I see you already say that in the next line :)